### PR TITLE
[WFLY-20005] Persistence/JPA subsystem should remove non-breaking use of ModuleIdentifier

### PR DESCRIPTION
**Pull request by scottmarlow**
https://issues.redhat.com/browse/WFLY-20005

### CODE REVIEW

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPADependencyProcessor.java [hunk 1]
```diff
@@ -229,10 +217,12 @@ private int loadPersistenceUnits(final ModuleSpecification moduleSpecification,
...
                     if (providerModuleName != null) {
-                        addOptionalDependency(moduleSpecification, moduleLoader, deploymentUnit, ModuleIdentifier.fromString(providerModuleName));
-                        ROOT_LOGGER.debugf("%s is configured to use persistence provider '%s', adding an optional dependency on module '%s'",
-                                pu.getPersistenceUnitName(), pu.getPersistenceProviderClassName(), providerModuleName);
+                        String moduleIdentifier = providerModuleName;
+                        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, moduleIdentifier, true, false, false, false));
+                        ROOT_LOGGER.debugf("added %s dependency to %s", moduleIdentifier, deploymentUnit.getName());
                     }
+                    ROOT_LOGGER.debugf("%s is configured to use persistence provider '%s', adding an optional dependency on module '%s'",
+                            pu.getPersistenceUnitName(), pu.getPersistenceProviderClassName(), providerModuleName);
```
Comment (yrodiere) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPADependencyProcessor.java:225]: This probably shouldn't be displayed if `providerModuleName` is `null`?

**Review by yrodiere — APPROVED** [review]
See also https://hibernate.zulipchat.com/#narrow/channel/132092-hibernate-search-dev/topic/WFLY-20005

As far as I can see this is semantically equivalent to what we had before, apart maybe from a few debug logs. So, +1 from me.

Thanks for taking care of it!

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPADependencyProcessor.java (hunk 1 above)
Comment (scottmarlow) [reply jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPADependencyProcessor.java]: Addressed ^

**Review by bstansberry — CHANGES_REQUESTED** [review]
I added the core-upgrade-needed label because this needs to use something like https://github.com/wildfly/wildfly-core/pull/6270 to handle module name strings passed in from outside.

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/messages/JpaLogger.java [hunk 2]
```diff
@@ -298,18 +298,6 @@ public interface JpaLogger extends BasicLogger {
...
-//     * for the adapter, represented by the {@code name} parameter.
-//     *
-//     * @param cause    the cause of the error.
-//     * @param moduleId the module id that was attempting to be loaded.
-//     * @param name     the name of the adapter.
-//     * @return a {@link RuntimeException} for the error.
-//     */
-    //@Message(id = 26, value = "Could not load module %s to add %s adapter to deployment")
-    //RuntimeException cannotLoadModule(@Cause Throwable cause, ModuleIdentifier moduleId, String name);
-
```
[the remark this answers is not in the document]
Comment (scottmarlow) [reply jpa/subsystem/src/main/java/org/jboss/as/jpa/messages/JpaLogger.java]: Restored.  I removed due to api change but probably okay to leave this unchanged.

[the remark this answers is not in the document]
Comment (bstansberry) [reply jpa/subsystem/src/main/java/org/jboss/as/jpa/messages/JpaLogger.java]: Thanks. There's no expectation that these methods would ever be uncommented. They are just a form of documentation of the how the message ids were used.

**Comment by bstansberry** [comment]
@scottmarlow The https://github.com/wildfly/wildfly-core/pull/6270 API is now available.

**Comment by scottmarlow** [comment]
> @scottmarlow The [wildfly/wildfly-core#6270](https://github.com/wildfly/wildfly-core/pull/6270) API is now available.

Thanks @bstansberry  for the new API, very helpful!  

Please check https://github.com/wildfly/wildfly/pull/18439/commits/00e6180836ff5b169f242d9114a9728fe9fd9ab6 + https://github.com/wildfly/wildfly/pull/18439/commits/2bcf75bb2663c24cd3063d219eaae2b7029d0ffa  to see updates to use the new https://github.com/wildfly/wildfly-core/pull/6270 API.

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/HibernateSearchProcessor.java [hunk 3]
```diff
@@ -93,18 +81,20 @@ else if (!providerModule.equals(searchMapperModuleName)) {
 
         // use Search module name specified in persistence unit definition
         if (searchMapperModuleName != null && !IGNORE.equals(searchMapperModuleName)) {
-            ModuleIdentifier moduleIdentifier = ModuleIdentifier.fromString(searchMapperModuleName);
-            moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, moduleIdentifier, false, true, true, false));
-            ROOT_LOGGER.debugf("added %s dependency to %s", moduleIdentifier, deploymentUnit.getName());
+            moduleSpecification.addSystemDependency(ModuleDependency.Builder.of(moduleLoader, searchMapperModuleName).
```
Comment (bstansberry) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/HibernateSearchProcessor.java:84]: I believe the searchMapperModuleName var has not been canonicalized. It holds a value that was read from PersistenceUnitMetadata.getProperties().getProperty(Configuration.HIBERNATE_SEARCH_MODULE) and I don't expect that value to have been canonicalized.

I was going to suggest a fix, but the only use of this is to pass it into ModuleDependency.Builder, and that handles the canonicalization.

**Review by bstansberry — APPROVED** [review]

**Comment by bstansberry** [comment]
Thanks @scottmarlow and @yrodiere

**Pull request merged**